src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java

```diff
@@ -0,0 +1,364 @@
+	{
+		doFtpDownload(null, null);
+	}
+	
+	protected void doFtpDownload(String user, String password) throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		user = user == null || user.trim().equals("") ? "anonymous" : user;
+		password = password == null || password.trim().equals("") ? "" : password;
+		FTPClient client = new FTPClient();
+		
+		client.connect(this.uri.getHost());
+		if (this.passiveFTP)
+		{
+			client.enterLocalPassiveMode(); //PASSIVE mode works better when inside a docker container.
+		}
+		client.login(user, password);
+		logger.debug("connect/login reply code: {}",client.getReplyCode());
+		client.setFileType(FTP.BINARY_FILE_TYPE);
+		client.setFileTransferMode(FTP.COMPRESSED_TRANSFER_MODE);
+		try
+		{
+			try(InputStream inStream = client.retrieveFileStream(this.uri.getPath()))
+			{
+				if (inStream != null)
+				{
+					writeInputStreamToFile(inStream);
+				}
+				else
+				{
+					logger.error("No data returned from server for {}", this.uri.toString());
+				}
+			}
+		}
+		catch (IOException e)
+		{
+			logger.error("Error while retrieving the file: {}",e.getMessage());
+			e.printStackTrace();
+			throw new IOException(e);
+		}
+		//Should probably have more/better reply-code checks.
+		logger.debug("retreive file reply code: {}",client.getReplyCode());
+		if (client.getReplyString().matches("^5\\d\\d.*") || (client.getReplyCode() >= 500 && client.getReplyCode() < 600) )
+		{
+			String errorString = "5xx reply code detected (" + client.getReplyCode() + "), reply string is: "+client.getReplyString();
+			logger.error(errorString);
+			throw new Exception(errorString);
+		}
+		client.logout();
+		client.disconnect();
+	}
+
+	protected void writeInputStreamToFile(InputStream inStream) throws IOException, FileNotFoundException
+	{
+		try(ByteArrayOutputStream baos = new ByteArrayOutputStream())
+		{
+			int b = inStream.read();
+			while (b!=-1)
+			{
+				baos.write(b);
+				b = inStream.read();
+			}
+	
+			try (FileOutputStream file = new FileOutputStream(this.destination))
+			{
+				baos.writeTo(file);
+				file.flush();
+			}
+		}
+	}
+
+	
+	protected void doHttpDownload(Path path) throws HttpHostConnectException, IOException, Exception
+	{
+		this.doHttpDownload(path, HttpClientContext.create());
+	}
+	
+	protected void doHttpDownload(Path path, HttpClientContext context) throws Exception, HttpHostConnectException, IOException
+	{
+		HttpGet get = new HttpGet(this.uri);
+		//Need to multiply by 1000 because timeouts are in milliseconds.
+		RequestConfig config = RequestConfig.copy(RequestConfig.DEFAULT)
+											.setConnectTimeout(1000 * (int)this.timeout.getSeconds())
+											.setSocketTimeout(1000 * (int)this.timeout.getSeconds())
+											.setConnectionRequestTimeout(1000 * (int)this.timeout.getSeconds()).build();
+		
+		get.setConfig(config);
+		
+		int retries = this.numRetries;
+		boolean done = retries + 1 <= 0;
```

Comment:
Why not `retries < 0`?

Reply:
It's been a while... I think it had something to do with the message printing out something like `...but will retry 0 more time(s)` right before it did the last retry (and stopping when it hit -1). It makes more sense if it says "will retry 1 more time" before that last retry. Are there other ways to do this? Yes, this is just one way.

Reply:
Yes, but as a new reader I find it confusing that retries can potentially be -1.  I think it's better to refactor so retries can't be below 0:

```
boolean done;
do {
    ...
    // If retry not needed
    done = true;
    // If retry needed
    if (retries == 0) {
        done = true;
    } else {
        retries--;
    }
} while (!done);